File: sdk/go/synaptic.go
package heartlight

import "math"

const SynapticKernelVersion = "1.0"

type SynapticConfig struct {
	LearningRate float64 `json:"learning_rate"`
	Decay        float64 `json:"decay"`
	TraceDecay   float64 `json:"trace_decay"`
	MinWeight    float64 `json:"min_weight"`
	MaxWeight    float64 `json:"max_weight"`
}

type SynapticState struct {
	Weight      float64 `json:"weight"`
	Eligibility float64 `json:"eligibility"`
}

type SynapticInput struct {
	Pre    float64 `json:"pre"`
	Post   float64 `json:"post"`
	Reward float64 `json:"reward"`
}

func DefaultSynapticConfig() SynapticConfig {
	return SynapticConfig{LearningRate: 0.08, Decay: 0.002, TraceDecay: 0.9, MinWeight: -1.0, MaxWeight: 1.0}
}

func SynapticStep(state SynapticState, input SynapticInput, config SynapticConfig) SynapticState {
	values := []float64{config.LearningRate, config.Decay, config.TraceDecay, config.MinWeight, config.MaxWeight, state.Weight, state.Eligibility, input.Pre, input.Post, input.Reward}
	for _, value := range values {
		if math.IsNaN(value) || math.IsInf(value, 0) {
			panic("HEARTLIGHT synaptic values must be finite")
		}
	}
	if config.LearningRate < 0 || config.Decay < 0 || config.Decay > 1 || config.TraceDecay < 0 || config.TraceDecay > 1 || config.MinWeight > config.MaxWeight {
		panic("invalid HEARTLIGHT synaptic config")
	}
	eligibility := config.TraceDecay*state.Eligibility + input.Pre*input.Post
	weight := (1-config.Decay)*state.Weight + config.LearningRate*input.Reward*eligibility
	weight = math.Max(config.MinWeight, math.Min(config.MaxWeight, weight))
	return SynapticState{Weight: weight, Eligibility: eligibility}
}

File: sdk/typescript/src/synaptic.ts
export const SYNAPTIC_KERNEL_VERSION = "1.0" as const;

export type SynapticConfig = {
  learning_rate: number;
  decay: number;
  trace_decay: number;
  min_weight: number;
  max_weight: number;
};

export type SynapticState = { weight: number; eligibility: number };
export type SynapticInput = { pre: number; post: number; reward: number };

export const defaultSynapticConfig = (): SynapticConfig => ({
  learning_rate: 0.08,
  decay: 0.002,
  trace_decay: 0.9,
  min_weight: -1.0,
  max_weight: 1.0,
});

export function synapticStep(
  state: SynapticState,
  input: SynapticInput,
  config: SynapticConfig = defaultSynapticConfig(),
): SynapticState {
  const values = [
    config.learning_rate,
    config.decay,
    config.trace_decay,
    config.min_weight,
    config.max_weight,
    state.weight,
    state.eligibility,
    input.pre,
    input.post,
    input.reward,
  ];
  if (!values.every(Number.isFinite)) throw new TypeError("HEARTLIGHT synaptic values must be finite");
  if (
    config.learning_rate < 0 ||
    config.decay < 0 || config.decay > 1 ||
    config.trace_decay < 0 || config.trace_decay > 1 ||
    config.min_weight > config.max_weight
  ) throw new RangeError("invalid HEARTLIGHT synaptic config");

  const eligibility = config.trace_decay * state.eligibility + input.pre * input.post;
  const rawWeight = (1 - config.decay) * state.weight + config.learning_rate * input.reward * eligibility;
  const weight = Math.max(config.min_weight, Math.min(config.max_weight, rawWeight));
  return { weight, eligibility };
}
